File: Shopify/java/RobotGrid.java
import java.util.*;

/**
 * Shopify Pair Programming - Robot / Mars Rover on a Grid
 * Confirmed: Aug 2024, Apr 2025 (JoinTaro).
 *
 * Java 11.  Run:  javac RobotGrid.java && java RobotGrid
 *
 * Layer 1: single robot, L/R/M commands, report (x, y, dir)
 * Layer 2: bounded grid + obstacles, blocked moves
 * Layer 3: serialize / deserialize state
 * Layer 4 (staff, discussion only): HTTP API, concurrency, failure modes
 */
public class RobotGrid {

    /** Clockwise order (N,E,S,W) makes turning trivial via ordinal arithmetic. */
    enum Direction {
        NORTH(0, 1), EAST(1, 0), SOUTH(0, -1), WEST(-1, 0);
        final int dx, dy;
        Direction(int dx, int dy) { this.dx = dx; this.dy = dy; }
        Direction left()  { return values()[(ordinal() + 3) % 4]; }
        Direction right() { return values()[(ordinal() + 1) % 4]; }
    }

    /** Immutable value type. (Java 17+ would make this a record.) */
    static final class Position {
        private final int x, y;
        private final Direction dir;
        Position(int x, int y, Direction dir) { this.x = x; this.y = y; this.dir = dir; }
        int x() { return x; }
        int y() { return y; }
        Direction dir() { return dir; }
        Position move()  { return new Position(x + dir.dx, y + dir.dy, dir); }
        Position left()  { return new Position(x, y, dir.left()); }
        Position right() { return new Position(x, y, dir.right()); }
        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Position)) return false;
            Position p = (Position) o;
            return x == p.x && y == p.y && dir == p.dir;
        }
        @Override public int hashCode() { return Objects.hash(x, y, dir); }
        @Override public String toString() { return "(" + x + "," + y + "," + dir + ")"; }
    }

    // ---- Layer 1: unbounded movement ----
    static Position execute(String commands) {
        Position p = new Position(0, 0, Direction.NORTH);
        for (char c : commands.toCharArray()) {
            switch (c) {
                case 'L': p = p.left();  break;
                case 'R': p = p.right(); break;
                case 'M': p = p.move();  break;
                default:  throw new IllegalArgumentException("Unknown command: " + c);
            }
        }
        return p;
    }

    // ---- Layer 2 + 3: bounded grid, obstacles, serialization ----
    static final class Cell {
        private final int x, y;
        Cell(int x, int y) { this.x = x; this.y = y; }
        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Cell)) return false;
            Cell c = (Cell) o;
            return x == c.x && y == c.y;
        }
        @Override public int hashCode() { return Objects.hash(x, y); }
    }

    static class Robot {
        private final int width, height;
        private final Set<Cell> obstacles = new HashSet<>();
        private Position pos = new Position(0, 0, Direction.NORTH);

        Robot(int width, int height) {
            if (width <= 0 || height <= 0)
                throw new IllegalArgumentException("Grid dimensions must be positive");
            this.width = width;
            this.height = height;
        }

        void addObstacle(int x, int y) { obstacles.add(new Cell(x, y)); }

        /** Returns false if a move was blocked by a wall or obstacle (stops there). */
        boolean execute(String commands) {
            for (char c : commands.toCharArray()) {
                switch (c) {
                    case 'L': pos = pos.left();  break;
                    case 'R': pos = pos.right(); break;
                    case 'M':
                        Position next = pos.move();
                        if (!inBounds(next) || obstacles.contains(new Cell(next.x(), next.y())))
                            return false;
                        pos = next;
                        break;
                    default: throw new IllegalArgumentException("Unknown command: " + c);
                }
            }
            return true;
        }

        Position position() { return pos; }

        private boolean inBounds(Position p) {
            return p.x() >= 0 && p.x() < width && p.y() >= 0 && p.y() < height;
        }

        String serialize() { return pos.x() + "," + pos.y() + "," + pos.dir(); }

        static Robot deserialize(String state, int width, int height) {
            String[] parts = state.split(",");
            Robot r = new Robot(width, height);
            r.pos = new Position(Integer.parseInt(parts[0]),
                                 Integer.parseInt(parts[1]),
                                 Direction.valueOf(parts[2]));
            return r;
        }
    }

    /* ---- Layer 4 talking points (no code) ----
     * HTTP API:  POST /robots, POST /robots/{id}/commands, GET /robots/{id}
     * 1000 robots: each robot is independent state -> shard by id (consistent hashing).
     * Idempotency: attach a sequence number to each command batch.
     * Failure modes: lost commands -> command log + replay (event sourcing);
     *                two robots same cell -> reservation/lock on the grid.
     */

    public static void main(String[] args) {
        // Layer 1
        Position p = execute("MMRMMRMM");
        check("Layer1", p.equals(new Position(2, 0, Direction.SOUTH)), p.toString());

        // Layer 2: blocked by an obstacle, position frozen at the wall
        Robot r = new Robot(5, 5);
        r.addObstacle(0, 2);
        boolean ok = r.execute("MM");
        check("Blocked", !ok && r.position().equals(new Position(0, 1, Direction.NORTH)),
              r.position().toString());

        // Layer 3: serialize round-trip
        Robot r2 = new Robot(10, 10);
        r2.execute("MRMM");
        Robot restored = Robot.deserialize(r2.serialize(), 10, 10);
        check("Serialize", restored.serialize().equals(r2.serialize()), restored.serialize());
    }

    static void check(String name, boolean pass, String actual) {
        System.out.printf("%-10s %s  (%s)%n", name, pass ? "PASS" : "FAIL", actual);
    }
}

File: Shopify/pair_programming/robot_grid.cpp
/*
 * Shopify Pair Programming - Most Frequently Reported Problem
 * Robot / Roomba on a Grid
 *
 * Problem (Layer 1):
 *   A robot starts at position (0,0) facing NORTH on a 10x10 grid.
 *   Commands: "L" = turn left, "R" = turn right, "M" = move forward 1 step.
 *   Return final (x, y, direction).
 *
 * Extensions reported in interviews:
 *   Layer 2: Add obstacles. Robot stops and reports collision.
 *   Layer 3: Multiple robots. Serialize/deserialize robot state.
 *   Layer 4 (staff): How would you expose this as an HTTP API?
 *              How would you handle 1000 concurrent robots?
 *              What are the failure modes?
 *
 * Shopify interview tip:
 *   Start with Layer 1 working and running. Volunteer Layer 2 before being asked.
 *   Use clean enums and structs — they probe readability.
 */

#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <stdexcept>

// ============================================================
// Layer 1: Basic robot movement on unbounded grid
// ============================================================

enum class Direction { NORTH, EAST, SOUTH, WEST };

std::string dirToStr(Direction d) {
    switch (d) {
        case Direction::NORTH: return "NORTH";
        case Direction::EAST:  return "EAST";
        case Direction::SOUTH: return "SOUTH";
        case Direction::WEST:  return "WEST";
    }
    return "";
}

Direction turnLeft(Direction d) {
    // NORTH -> WEST -> SOUTH -> EAST -> NORTH
    return static_cast<Direction>((static_cast<int>(d) + 3) % 4);
}

Direction turnRight(Direction d) {
    return static_cast<Direction>((static_cast<int>(d) + 1) % 4);
}

struct Position {
    int x, y;
    Direction dir;
};

Position move(Position pos) {
    switch (pos.dir) {
        case Direction::NORTH: pos.y++; break;
        case Direction::EAST:  pos.x++; break;
        case Direction::SOUTH: pos.y--; break;
        case Direction::WEST:  pos.x--; break;
    }
    return pos;
}

Position executeCommands(const std::string& commands) {
    Position pos = {0, 0, Direction::NORTH};
    for (char c : commands) {
        if (c == 'L')      pos.dir = turnLeft(pos.dir);
        else if (c == 'R') pos.dir = turnRight(pos.dir);
        else if (c == 'M') pos = move(pos);
        else throw std::invalid_argument(std::string("Unknown command: ") + c);
    }
    return pos;
}

// ============================================================
// Layer 2: Bounded grid + obstacles
// ============================================================

class Robot {
public:
    Robot(int gridW, int gridH) : gridW(gridW), gridH(gridH), pos({0, 0, Direction::NORTH}) {}

    void addObstacle(int x, int y) { obstacles.insert({x, y}); }

    // Returns false if blocked by obstacle or boundary
    bool execute(const std::string& commands) {
        for (char c : commands) {
            if (c == 'L')      pos.dir = turnLeft(pos.dir);
            else if (c == 'R') pos.dir = turnRight(pos.dir);
            else if (c == 'M') {
                Position next = move(pos);
                if (!inBounds(next) || obstacles.count({next.x, next.y})) {
                    return false; // blocked
                }
                pos = next;
            }
        }
        return true;
    }

    Position getPosition() const { return pos; }

    // Layer 3: Serialize / deserialize state
    std::string serialize() const {
        return std::to_string(pos.x) + "," + std::to_string(pos.y) + "," + dirToStr(pos.dir);
    }

    static Robot deserialize(const std::string& state, int gridW, int gridH) {
        // parse "x,y,DIR"
        Robot r(gridW, gridH);
        size_t p1 = state.find(',');
        size_t p2 = state.find(',', p1 + 1);
        r.pos.x = std::stoi(state.substr(0, p1));
        r.pos.y = std::stoi(state.substr(p1 + 1, p2 - p1 - 1));
        std::string d = state.substr(p2 + 1);
        if (d == "NORTH")      r.pos.dir = Direction::NORTH;
        else if (d == "EAST")  r.pos.dir = Direction::EAST;
        else if (d == "SOUTH") r.pos.dir = Direction::SOUTH;
        else                   r.pos.dir = Direction::WEST;
        return r;
    }

private:
    int gridW, gridH;
    Position pos;
    std::set<std::pair<int,int>> obstacles;

    bool inBounds(const Position& p) const {
        return p.x >= 0 && p.x < gridW && p.y >= 0 && p.y < gridH;
    }
};

// ============================================================
// Staff Layer 4 Discussion Points (not code — talking points)
// ============================================================
/*
 * Q: How would you expose this as an HTTP API?
 *    - POST /robots          -> create robot, return id
 *    - POST /robots/:id/commands  -> body: {"commands": "MMLMR"}
 *    - GET  /robots/:id      -> returns state
 *    State stored in Redis (fast reads), persisted to DB async.
 *
 * Q: 1000 concurrent robots?
 *    - Each robot is independent state — horizontally scalable.
 *    - Robot ID routes to a shard (consistent hashing).
 *    - Command processing is idempotent if we add sequence numbers.
 *
 * Q: Failure modes?
 *    - Lost commands: use command log + replay (event sourcing pattern).
 *    - Stale state: optimistic locking on position updates.
 *    - What if two robots collide? Need a global grid lock or reservation system.
 */

// ============================================================
// Tests (run these during the interview to show TDD mindset)
// ============================================================

void runTests() {
    // Layer 1 tests
    {
        auto pos = executeCommands("MMRMMRMM");
        // Start NORTH, move 2 (y=2), turn right (EAST), move 2 (x=2),
        // turn right (SOUTH), move 2 (y=0)
        std::cout << "Test 1: " << pos.x << "," << pos.y << "," << dirToStr(pos.dir) << std::endl;
        // Expected: 2,0,SOUTH
    }
    {
        auto pos = executeCommands("LMLMLMLMM");
        std::cout << "Test 2: " << pos.x << "," << pos.y << "," << dirToStr(pos.dir) << std::endl;
        // 0 moves net in x, 3 in y, facing NORTH — verify by hand
    }

    // Layer 2 tests
    {
        Robot r(5, 5);
        r.addObstacle(0, 2);
        bool ok = r.execute("MM"); // would hit obstacle at (0,2)
        auto pos = r.getPosition();
        std::cout << "Test 3 (blocked): ok=" << ok << " pos=" << pos.x << "," << pos.y << std::endl;
        // Expected: ok=false, pos=(0,1)
    }

    // Layer 3 serialize/deserialize
    {
        Robot r(10, 10);
        r.execute("MRMM");
        std::string s = r.serialize();
        Robot r2 = Robot::deserialize(s, 10, 10);
        std::cout << "Test 4 (serialize): " << r2.serialize() << std::endl;
        // Should match r.serialize()
    }
}

int main() {
    runTests();
    return 0;
}
